### ISSUE TITLE
reasoning-budget: match llama.cpp's graceful block close (forced message, UTF-8 hold, re-arm, multi-token markers)

### ISSUE THREAD

**Issue by ehartford** [QuixiAI/SlimServe#20]
## Summary

Our `thinking_token_budget` enforcement (PR #13) and llama.cpp's
`--reasoning-budget` agree on the core policy — neither injects a soft
"wrap up" nudge partway through, and both force the reasoning block shut
only once the budget reaches zero. Where they differ is *how* they close
it. llama.cpp closes the block gracefully; we sever it.

This issue tracks bringing our implementation to parity on four points,
using `~/llama.cpp` `common/reasoning-budget.{h,cpp}` as the reference.

Deliberately **not** in scope: a percentage-threshold nudge at ~85–90%.
Neither implementation has one, and forcing logits cannot express "start
concluding" — the sampler stops consulting the model entirely. That is a
separate design question; this issue is about matching a known-good
reference, not inventing past it.

## Reference

- `common/reasoning-budget.h` — 5-state machine and API
- `common/reasoning-budget.cpp` — state transitions, matcher
- `common/sampling.cpp:311-317` — construction
- `common/arg.cpp:3728-3742` — `--reasoning-budget`, `--reasoning-budget-message`
- `tests/test-reasoning-budget.cpp` — 505 lines of coverage

Ours:

- `vllm/v1/worker/gpu/sample/thinking_budget.py` — V2 GPU enforcement
- `vllm/v1/sample/thinking_budget_state.py` — V1 logits-processor path

### 1. Forced sequence, not a single token

llama.cpp forces `forced_tokens` = *message + first end tag*, emitted
token-by-token across successive steps in `REASONING_BUDGET_FORCING`.
`--reasoning-budget-message` (default empty) is injected **before** the
end tag, so a truncated block reads:

```
...as I was saying the second case  [Time's up, answering now.] </think>
```

We can only ever force one token. `thinking_budget.py:186-190` masks the
whole vocabulary to `-inf` and leaves `end_id` — so the chain of thought
stops mid-word and the answer phase begins from a severed clause.

This is the gap that actually shows up in output quality, and it is the
one worth doing first. Note it is still not a nudge: the text is forced,
the model cannot react to it. It buys a coherent handoff, not
self-directed convergence.

### 2. No UTF-8 boundary hold

llama.cpp has a dedicated `REASONING_BUDGET_WAITING_UTF8` state: when the
budget expires mid-multi-byte character, generation continues until the
codepoint closes, *then* forcing begins.

We force at the exact step `req_remaining <= 0`
(`thinking_budget.py:166`), which can cut a partial UTF-8 sequence and
emit a broken character. This is a correctness bug for CJK and emoji, not
a polish item.

### 3. Single-token markers only

llama.cpp matches multi-token start/end tag *sequences* via Aho-Corasick
(`token_matcher`, `common_trie`), supports several end tags, and reports
which one matched (`common_reasoning_budget_get_end_match`).

We compare against scalar `self.start_id` / `self.end_id`. To be fair to
the current code this is **not** silent: `reasoning_markers_are_single_token`
gates construction, and `input_processor.py:126` raises a clear
`ValueError` naming the cause and the `VLLM_USE_V2_MODEL_RUNNER=0`
workaround. So this is a declared limitation, and the lowest priority
here — but it does mean any model whose markers tokenize to multiple
tokens has no V2 budget support at all.

### 4. No re-arm on a second reasoning block

llama.cpp returns `DONE -> COUNTING` with a **fresh** budget when a new
start tag appears; the code comments that a response may contain several
reasoning blocks and each should get its own window.

Ours sets `remaining` once at admission and only ever decrements it
(`thinking_budget.py:211`). `is_start` flips `in_think` (lines 178, 212)
but never resets the counter. So a second `<think>` block inherits the
first's leftover budget — and if the first exhausted it, the second is
forced shut on its first token. Verified by reading; no test covers it.

## Constraint on the fix

llama.cpp's is a per-slot CPU sampler with free branching. Ours is
batched GPU tensor ops deliberately kept **shape-static** — the comment at
`thinking_budget.py:185-187` records that boolean advanced indexing gives
data-dependent shapes that drain the GPU queue per step on MPS.

So the forced sequence must be implemented as a per-request cursor into a
padded token tensor, advanced with masked writes, not as a variable-length
Python loop. Same for the UTF-8 hold: it needs a per-request "held" flag,
not an early return.

Both runners need the change (V1 `thinking_budget_state.py` and V2
`thinking_budget.py`), and the V2 path additionally has to work inside the
speculative verify span, where consumption accumulates per request across
rows rather than per row.

## Suggested order

1. Forced sequence + budget message (biggest quality win)
2. UTF-8 boundary hold (correctness)
3. Re-arm on new start tag (correctness, narrow)
4. Multi-token markers (removes a declared restriction)

## Acceptance

- A budget-exhausted block ends with the configured message then the end
  tag, on both runners, with and without speculation.
- Budget expiry mid-multi-byte character emits no broken codepoint.
- A second reasoning block in one response gets a fresh budget.
- Port the relevant cases from `tests/test-reasoning-budget.cpp`.
- Exact-token parity run on a SlimServe profile confirms no regression in
  the shape-static sampling path.
